sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -4600,3 +4600,133 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
+       ["b","d","c","a","d"]
+
+  """,
+  since = "3.4.0",
+  group = "array_funcs")
+case class ArrayAppend(left: Expression, right: Expression)
+  extends BinaryExpression
+  with ImplicitCastInputTypes
+  with ComplexTypeMergingExpression
+  with QueryErrorsBase {
+  override def prettyName: String = "array_append"
+
+  override def inputTypes: Seq[AbstractDataType] = {
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, hasNull), e2) =>
+        TypeCoercion.findTightestCommonType(e1, e2) match {
+          case Some(dt) => Seq(ArrayType(dt, hasNull), dt)
+          case _ => Seq.empty
+        }
+      case _ => Seq.empty
+    }
+  }
+
+  override def eval(input: InternalRow): Any = {
+    val value1 = left.eval(input)
+    val value2 = right.eval(input)
+    if (value1 == null) {
+      null
+    } else {
+      nullSafeEval(value1, value2)
+    }
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, _), e2) => if (e1.sameType(e2)) {
+        TypeCheckResult.TypeCheckSuccess
+      }
+      else {
+        DataTypeMismatch(
+          errorSubClass = "ARRAY_FUNCTION_DIFF_TYPES",
+          messageParameters = Map(
+            "functionName" -> toSQLId(prettyName),
+            "leftType" -> toSQLType(left.dataType),
+            "rightType" -> toSQLType(right.dataType),
+            "dataType" -> toSQLType(ArrayType)
+          ))
+      }
+      case _ =>
+        DataTypeMismatch(
+          errorSubClass = "UNEXPECTED_INPUT_TYPE",
+          messageParameters = Map(
+            "paramIndex" -> "0",
+            "requiredType" -> toSQLType(ArrayType),
+            "inputSql" -> toSQLExpr(left),
+            "inputType" -> toSQLType(left.dataType)
+          )
+        )
+    }
+  }
+
+  protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression): ArrayAppend =
+    copy(left = newLeft, right = newRight)
+
+  override protected def nullSafeEval(input1: Any, input2: Any): Any = {
+    val arrayData = input1.asInstanceOf[ArrayData]
+    val arrayElementType = dataType.asInstanceOf[ArrayType].elementType
+    val elementData = input2
+    val numberOfElements = arrayData.numElements() + 1
+    if (numberOfElements > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+      throw QueryExecutionErrors.concatArraysWithElementsExceedLimitError(numberOfElements)
+    }
+    val finalData = new Array[Any](numberOfElements)
+    arrayData.foreach(arrayElementType, finalData.update)
+    finalData.update(numberOfElements - 1, elementData)
+    new GenericArrayData(finalData)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+
+    val f = (left: String, right: String) => {
+      val expr = ctx.addReferenceObj("arraysAppendExpr", this)
+      s"${ev.value} = (ArrayData)$expr.nullSafeEval($left, $right);"
+    }
+
+    val leftGen = left.genCode(ctx)
+    val rightGen = right.genCode(ctx)
```

Comment:
@LuciferYang One more thing I noticed, if we are having a null int value, it is getting translated to -1. It is picking up the default value for Integer primitive type, in the JavaCode. I think we should override this behaviour right ?

Reply:
Since it will automatically translate all null elements to default values for primitive types atleast.

Reply:
To give you a sense of the problem
```
    val df9 = spark.createDataFrame(
      spark.sparkContext.parallelize(
        Seq(Row(Array[Int](1, 2, 3), null))),
      StructType(List(
        StructField("a", ArrayType.apply(IntegerType), true),
        StructField("b", IntegerType, true)
      ))
    )
    df9.show
    df9.selectExpr("array_append(a, b)").printSchema()
    checkAnswer(df9.selectExpr("array_append(a, b)"),
      Seq(Row(Seq(1, 2, 3, null)))
    )

```

Output
```
== Results ==
!== Correct Answer - 1 ==   == Spark Answer - 1 ==
!struct<>                   struct<array_append(a, b):array<int>>
![List(1, 2, 3, null)]      [ArrayBuffer(1, 2, 3, -1)]
    
```

Reply:
> Integer

Will `null` in the src Array also be translated to `-1`?

Reply:
Nope the elements inside the src Array are not being translated to -1. 

```
    
    val df10 = spark.createDataFrame(
      spark.sparkContext.parallelize(
        Seq(Row(Seq[Integer](1, 2, 3, null), null))),
      StructType(List(
        StructField("a", ArrayType.apply(IntegerType), true),
        StructField("b", IntegerType, true)
      ))
    )

    df10.selectExpr("array_append(a, b)").printSchema()
    checkAnswer(df10.selectExpr("array_append(a, b)"),
      Seq(Row(Seq(1, 2, 3, null, null)))
    )
```
Output
```
== Results ==
!== Correct Answer - 1 ==      == Spark Answer - 1 ==
!struct<>                      struct<array_append(a, b):array<int>>
![List(1, 2, 3, null, null)]   [ArrayBuffer(1, 2, 3, null, -1)]
    
```

Reply:
Personally,  I think we should `override this behaviour` to append `null`

Reply:
WDYT @zhengruifeng ?

Reply:
I was try to refer to other higher order functions which accept array & element as inputs:

```
scala> spark.sql("""SELECT a,b,array_contains(a,b),array_position(a,b),array_remove(a,b) FROM VALUES (ARRAY(1, NULL, 3), 1), (ARRAY(1, NULL, 3), NULL), (NULL, 1), (NULL, NULL) AS tab(a, b)""").show
+------------+----+--------------------+--------------------+------------------+
|           a|   b|array_contains(a, b)|array_position(a, b)|array_remove(a, b)|
+------------+----+--------------------+--------------------+------------------+
|[1, null, 3]|   1|                true|                   1|         [null, 3]|
|[1, null, 3]|null|                null|                null|              null|
|        null|   1|                null|                null|              null|
|        null|null|                null|                null|              null|
+------------+----+--------------------+--------------------+------------------+
```

It seems the expected output should be `NULL`, if one of the two input columns is `NULL`.

@HyukjinKwon  @cloud-fan

Reply:
Then it would not be consistent with snowflake's array_append function.

Reply:
Hey- similarly saw this issue for implementing ArrayInsert; from a consistency perspective, I think we should strive to override the behaviour personally- for example, the below is current behaviour:
>> F.array_union(F.array(F.lit(None)), F.array(F.lit(1), F.lit(2)))
>> [null, 1, 2]

Not sure how an override would work in scala yet, I think we'd need to use an optional

Reply:
I think its okay if we are not consistent with the behaviour we have in snowflake and are more consistent to the existing high order functions in spark itself. In that we won't need to override the eval function also, we won't need to write the codeGen function completely. We will just be able to create nullSafeCodeGen function and call the function we need to call.

Reply:
@beliefer Please refer to this thread. :)

Reply:
@infoankitp Sorry for the late reply.
Thanks very much for pointing out the issue of NULL handling. We had offline discussions with some SQL experts and decided to adopt snowflake's method of NULL handling which is more reasonable compared with existing SparkSQL's method, that is:

1. if the array is NULL, returns NULL;
2. if the array is NOT NULL and the element is NULL, append this NULL into the array;

would you mind making such update? Thanks

Reply:
@zhengruifeng Thanks for notifying ! Let me make the changes in accordance to this.